### PR TITLE
Remove HOST_CALLS contract and related stuff

### LINKED ISSUE #100563: Test suite crash in new exception handling
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=627272
Build error leg or test failing: System.Text.RegularExpressions.Tests.WorkItemExecution
Pull request: https://github.com/dotnet/runtime/pull/100110

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "at System.Runtime.ExceptionServices.InternalCalls.<RhpSfiNext>g____PInvoke|1_0(System.Runtime.StackFrameIterator*, UInt32*, Boolean*, Boolean*)",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=627272
**Error message validated:** `[at System.Runtime.ExceptionServices.InternalCalls.<RhpSfiNext>g____PInvoke|1_0(System.Runtime.StackFrameIterator*, UInt32*, Boolean*, Boolean*)`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 4/2/2024 11:17:59 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[627272](https://dev.azure.com/dnceng-public/public/_build/results?buildId=627272)|dotnet/runtime|[System.Text.RegularExpressions.Tests.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=627272&view=ms.vss-test-web.build-test-results-tab&runId=15393574&resultId=209715)|dotnet/runtime#100110|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|1|1|

### PR DESCRIPTION
**Pull request by janvorli**
The HOST_CALLS contract is obsolete remainder from the .NET framework era, this change removes all of the use of that contract and the related code.

Also closes #100563

### CODE REVIEW

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp [hunk 1]
```diff
@@ -1921,11 +1921,6 @@ HRESULT GetFunctionInfoInternal(LPCBYTE ip, EECodeInfo * pCodeInfo)
         EE_THREAD_NOT_REQUIRED;
         CAN_TAKE_LOCK;
         CANNOT_RETAKE_LOCK;
-
-
-        // If this is called asynchronously (from a hijacked thread, as with F1), it must not re-enter the
-        // host (SQL).  Corners will be cut to ensure this is the case
-        if (ShouldAvoidHostCalls()) { HOST_NOCALLS; } else { HOST_CALLS; }
```
Comment (jkotas) [inline src/coreclr/vm/proftoeeinterfaceimpl.cpp:1928]: Delete `ShouldAvoidHostCalls` method as well?

**Comment by jkotas** [comment]
> Also closes https://github.com/dotnet/runtime/issues/100563

How is this closing this issue?

The HOST_CALLS are debug-only checks. The log attached to the issue has `Internal CLR error`
https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-100110-merge-2305133847734c678e/System.Text.RegularExpressions.Tests/1/console.9c55cc56.log?helixlogtype=result

**Comment by janvorli** [comment]
> How is this closing this issue?
> 
> The HOST_CALLS are debug-only checks. The log attached to the issue has `Internal CLR error`

The dump for that issue that I've downloaded using the runfo shows a crash due to accessing a garbage address down the call chain of `CONTRACT_ASSERT`:
```
 # Child-SP          RetAddr               Call Site
00 000000ff`e5c73990 00007ff8`f02a9c06     coreclr!common_strnlen_simd<0,1,unsigned char>+0x17d [minkernel\crts\ucrt\src\appcrt\string\strnlen.cpp @ 152] 
01 000000ff`e5c73b20 00007ff8`f02ab02d     coreclr!common_strnlen<0,unsigned char>+0x26 [minkernel\crts\ucrt\src\appcrt\string\strnlen.cpp @ 185] 
02 000000ff`e5c73b50 00007ff8`f02917b8     coreclr!strnlen+0x1d [minkernel\crts\ucrt\src\appcrt\string\strnlen.cpp @ 204] 
03 000000ff`e5c73b80 00007ff8`f02913e2     coreclr!__crt_stdio_output::output_processor<char,__crt_stdio_output::string_output_adapter<char>,__crt_stdio_output::format_validation_base<char,__crt_stdio_output::string_output_adapter<char> > >::type_case_s_compute_narrow_string_length+0x28 [minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h @ 2354] 
04 000000ff`e5c73bb0 00007ff8`f028c402     coreclr!__crt_stdio_output::output_processor<char,__crt_stdio_output::string_output_adapter<char>,__crt_stdio_output::format_validation_base<char,__crt_stdio_output::string_output_adapter<char> > >::type_case_s+0x102 [minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h @ 2340] 
05 000000ff`e5c73bf0 00007ff8`f028871f     coreclr!__crt_stdio_output::output_processor<char,__crt_stdio_output::string_output_adapter<char>,__crt_stdio_output::format_validation_base<char,__crt_stdio_output::string_output_adapter<char> > >::state_case_type+0x82 [minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h @ 2066] 
06 000000ff`e5c73c60 00007ff8`f027bef5     coreclr!__crt_stdio_output::output_processor<char,__crt_stdio_output::string_output_adapter<char>,__crt_stdio_output::format_validation_base<char,__crt_stdio_output::string_output_adapter<char> > >::process+0x34f [minkernel\crts\ucrt\inc\corecrt_internal_stdio_output.h @ 1704] 
07 000000ff`e5c73cd0 00007ff8`f027cc71     coreclr!common_vsprintf<__crt_stdio_output::format_validation_base,char>+0x2a5 [minkernel\crts\ucrt\src\appcrt\stdio\output.cpp @ 170] 
08 000000ff`e5c74200 00007ff8`f02942b9     coreclr!common_vsprintf_s<char>+0x201 [minkernel\crts\ucrt\src\appcrt\stdio\output.cpp @ 273] 
09 000000ff`e5c742a0 00007ff8`ef917cc4     coreclr!__stdio_common_vsprintf_s+0x69 [minkernel\crts\ucrt\src\appcrt\stdio\output.cpp @ 302] 
0a (Inline Function) --------`--------     coreclr!_vsprintf_s_l+0x24 [C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\stdio.h @ 1491] 
0b 000000ff`e5c74330 00007ff8`ef8e3937     coreclr!sprintf_s+0x44 [C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\stdio.h @ 1827] 
0c 000000ff`e5c74390 00007ff8`efca2ad6     coreclr!CONTRACT_ASSERT+0x237 [D:\a\_work\1\s\src\coreclr\inc\contract.inl @ 541] 
0d 000000ff`e5c77a60 00007ff8`f02388cb     coreclr!EEContract::DoChecks+0x386 [D:\a\_work\1\s\src\coreclr\vm\eecontract.cpp @ 177] 
0e 000000ff`e5c77af0 00007ff8`f0245c05     coreclr!ExecutionManager::ReaderLockHolder::ReaderLockHolder+0xdb [D:\a\_work\1\s\src\coreclr\vm\codeman.cpp @ 657] 
.
.
.
coreclr!common_strnlen_simd<0,1,unsigned char>+0x17d:
00007ff8`f02aa50d c5fe6f00        vmovdqu ymm0,ymmword ptr [rax] ds:00010119`00000000=??
0:009> dq @rax
00010119`00000000  ????????`????????

```

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:1928]: Done

**Review by jkotas — APPROVED** [review]

**Pull request merged**